### ISSUE TITLE
how to handle missingmember, or nulls

### ISSUE THREAD

**Issue by kolsmyr** [microsoft/fsharplu#48]
Hi

I am trying to find a jsonparser, besides chiron, that can actually handle deserializing errors. I have tried out FSharpLu which seems nice but with a few flaws, or it could be me ;). 

What I'm trying to do is to use "tryDeserialize<'T>"  but when trying to parse any json string which is missing a member, the parser gives following structure, while expecting Choice2Of2 error:
{"field1":null } or {"field2" : 0 }, where the last is of type int.

My point is that it doesnt seem to be able to handle nulls or missingmembers.

I have tried with both default, compact and the CamelCaseSerializer way of doing it.

Example comes here:

type CamelCaseSettings =
        static member settings =
            let s =
                JsonSerializerSettings(
                    NullValueHandling = NullValueHandling.Include,
                    MissingMemberHandling = MissingMemberHandling.Error,
                    ContractResolver = Serialization.CamelCasePropertyNamesContractResolver())
            s.Converters.Add(CompactUnionJsonConverter(true))
            s
        static member formatting = Formatting.Indented

        type CamelCaseSerializer = With<CamelCaseSettings>

    type C =
    | Ompa
    | Lompa

    type NumberSpecial = NumberSpecial of int

    type Ac ={
        number: NumberSpecial
        du: C
        testingOption: string option
        simpleInt : int
    }

    type Ab = {
        text: string
    }

    type B =
        | Ab of Ab
        | Ac of Ac

    type A = {
        id: System.Guid
        context: B
        name: string
    }
//des1 works fine
let des1 ="""{   "id":"f893e695-496d-4e30-8fc7-b2ff59725e6c",
        "context": {"ac": {
                            "number": {
                                        "numberSpecial":2
                            },
                    "du":"ompa",
                    "testingOption":"hejsvejs",
                    "simpleInt":4}},
        "name":"hola"
    }"""

//des2 does not throw error when obviously missing members...
//its output is: 
// Choice1Of2 {id = 00000000-0000-0000-0000-000000000000;
//           context = Ac {number = null;
//                         du = Ompa;
//                          testingOption = None;
//                           simpleInt = 0;};
//              name = "hola";}

let des2 =
        """{
                "context": {
                "ac": {
                    "du": "ompa"
                }
                },
                "name": "hola"
            }"""

Or have i totally missunderstood how to work with FSharpLu?

Best / Johan

**Comment by blumu** [comment]
@kolsmyr Thanks for reporting. An easy workaround is to add the property `JsonObject(ItemRequired=Required.Always)` to the serialized type. In your code example that would be:
```fs
[<JsonObject(ItemRequired=Required.Always)>] 
type A = {
    id: System.Guid
    context: B
    name: string
}
```
The deserializatin of `des2` would then fail:
```
val it : Choice<A,string> =
  Choice2Of2
    "An exception occurred during deserialization: Newtonsoft.Json.JsonSerializationException: Required property 'id' not found in JSON. Path '', line 8, position 5.
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EndProcessProperty(Object newObject, JsonReader reader, JsonObjectContract contract, Int32 initialDepth, JsonProperty property, PropertyPresence presence, Boolean setDefaultValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObjectUsingCreatorWithParameters(JsonReader reader, JsonObjectContract contract, JsonProperty containerProperty, ObjectConstructor`1 creator, String id)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader reader, JsonObjectContract objectContract, JsonProperty containerMember, JsonProperty containerProperty, String id, Boolean& createdFromNonDefaultCreator)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at <StartupCode$FSI_0067>.$FSI_0067.main@()"
```

**Comment by blumu** [comment]
@kolsmyr I've implemented it in #54 Please take a look and let me know if it addresses your problem.

**Comment by blumu** [comment]
@kolsmyr Closing this issue since I believe it's now fixed and released. Please feel free to reopen if you the implemented fix does not address your issue.
